rust/noosphere-storage/src/implementation/sqlite.rs

```diff
@@ -0,0 +1,354 @@
+        SqliteStore::new(&self.client, name).await
+    }
+}
+
+#[async_trait]
+impl Storage for SqliteStorage {
+    type BlockStore = SqliteStore;
+
+    type KeyValueStore = SqliteStore;
+
+    async fn get_block_store(&self, name: &str) -> anyhow::Result<Self::BlockStore> {
+        self.get_store(name).await
+    }
+
+    async fn get_key_value_store(&self, name: &str) -> anyhow::Result<Self::KeyValueStore> {
+        self.get_store(name).await
+    }
+}
+
+#[derive(Clone)]
+pub struct SqliteStore {
+    client: SqliteClient,
+    table_name: String,
+}
+
+impl SqliteStore {
+    async fn new(client: &SqliteClient, table_name: &str) -> anyhow::Result<Self> {
+        client.create(table_name).await?;
+        Ok(SqliteStore {
+            client: client.to_owned(),
+            table_name: table_name.to_owned(),
+        })
+    }
+
+    // Try to interpret key bytes as a Cid. If that fails, try to
+    // interpret bytes as UTF-8.
+    fn key_bytes_to_key(key: &[u8]) -> anyhow::Result<String> {
+        match cid::Cid::try_from(key) {
+            Ok(cid) => Ok(cid.to_string()),
+            Err(_) => match String::from_utf8(key.to_owned()) {
+                Ok(string) => Ok(string),
+                Err(_) => Err(anyhow::anyhow!("Could not serialize sqlite key.")),
+            },
+        }
+    }
+}
+
+#[async_trait]
+impl Store for SqliteStore {
+    async fn read(&self, key: &[u8]) -> anyhow::Result<Option<Vec<u8>>> {
+        self.client
+            .read(&self.table_name, SqliteStore::key_bytes_to_key(key)?)
+            .await
+    }
+
+    async fn write(&mut self, key: &[u8], bytes: &[u8]) -> anyhow::Result<Option<Vec<u8>>> {
+        self.client
+            .write(
+                &self.table_name,
+                SqliteStore::key_bytes_to_key(key)?,
+                bytes.to_owned(),
+            )
+            .await
+    }
+
+    async fn remove(&mut self, key: &[u8]) -> anyhow::Result<Option<Vec<u8>>> {
+        self.client
+            .remove(&self.table_name, SqliteStore::key_bytes_to_key(key)?)
+            .await
+    }
+
+    async fn flush(&self) -> anyhow::Result<()> {
+        self.client.flush(&self.table_name).await
+    }
+}
+
+mod sqlite {
+    use super::*;
+    use rusqlite::{params, Connection, OptionalExtension, Transaction};
+    use std::result::Result;
+
+    #[derive(Clone, Debug)]
+    enum SqliteCommand {
+        Create {
+            table_name: String,
+        },
+        Read {
+            table_name: String,
+            key: String,
+        },
+        Write {
+            table_name: String,
+            key: String,
+            value: Vec<u8>,
+        },
+        Remove {
+            table_name: String,
+            key: String,
+        },
+        Flush {
+            _table_name: String,
+        },
+    }
+
+    #[derive(Clone, Debug)]
+    enum SqliteResponse {
+        Value(Option<Vec<u8>>),
+        Empty,
+    }
+
+    #[derive(Debug)]
+    pub struct SqliteClient {
```

Comment:
I think we should compartmentalize the SQLite work until we can work through the implementation together (I'm still not sure why we are managing a separate thread to access it, for example).